packages/x-data-grid-premium/src/hooks/features/aggregation/useGridAggregation.ts

```diff
@@ -63,6 +63,7 @@ export const useGridAggregation = (
       const currentModel = gridAggregationModelSelector(apiRef);
       if (currentModel !== model) {
         apiRef.current.setState(mergeStateWithAggregationModel(model));
+        apiRef.current.updateColumns([]);
         apiRef.current.forceUpdate();
```

Comment:
Why is this required?

While you're touching this, you could delete `.forceUpdate()`, it's a legacy function that should be removed.

Reply:
After some further debugging, it's not required as such, but there's a bug somehwere:
* Removing triggers `checkAggregationRulesDiff` -> triggers `hydrateColumns` -> updates columns (but aggregation is not cleared for whatever reason) -> triggers `viewportInnerSizeChange` -> triggers `handleGridSizeChange` (which often is unnecessary, unless there are flex columns) where aggregation is actually cleared.

So, it seems that `hydrateColumns` is probably ran before state has resolved.

The right fix then is to replace:
```typescript
useGridApiOptionHandler(apiRef, 'aggregationModelChange', checkAggregationRulesDiff);
```

With something along the lines of (unless there's a better internal way to chain this):
```
const aggregationModel = gridAggregationModelSelector(apiRef);
useEnhancedEffect(checkAggregationRulesDiff, [checkAggregationRulesDiff, aggregationModel]);
```

Reply:
Root cause is that these two get triggered out of order:
```typescript
useGridApiOptionHandler(apiRef, 'aggregationModelChange', checkAggregationRulesDiff);

React.useEffect(() => {
    if (props.aggregationModel !== undefined) {
      apiRef.current.setAggregationModel(props.aggregationModel);
    }
  }, [apiRef, props.aggregationModel]);
```

As soon as the prop changes, `checkAggregationRulesDiff` is triggered, ahead of `setAggregationModel`. And it currently only works, because column update always triggers two state updates to columns (which in most cases is not needed), which triggers a second hydration from aggregation, only at which point the aggregation state has been updated.

Reply:
cc @flaviendelangle, any thoughts? 🙏 Seems like you worked on this feature.

Reply:
I built the feature 2.5 years ago so my understanding of it might be rusted :laughing: 

> As soon as the prop changes, checkAggregationRulesDiff is triggered

That surprises me.
`checkAggregationRulesDiff` is triggered by the `aggregationModelChange` event, which is fired in the `setState` method, which is called by `setAggregationModel`

So basically:

`props.aggregationModel` changes => `setState` is called => ` state.aggregation.model` is updated and `aggregationModelChange` is fired => `checkAggregationRulesDiff` is called

Maybe we have a problem during the initialization though.

The problem with firing updated like `checkAggregationRulesDiff` in effects is that you might end up with inconsistent states and flickering, but I don't know if that's the case for this specific feature.

Reply:
There problem seems to be on initialization indeed. Not sure if it has any real-world bearing, but on first prop change, the values are as such:
```
aggregationRules {}
rulesOnLastColumnHydration {}
aggregationRules {}
areAggregationRulesEqual(rulesOnLastColumnHydration, aggregationRules)
```

`rulesOnLastColumnHydration` is empty, even though columns have rules applied as can be seen from the test result:
```
1) <DataGridPremium /> - Aggregation
       Setting aggregation model
         prop: aggregationModel
           should correctly restore the column when changing from aggregated to non-aggregated:

      AssertionError: expected 'idmax' to equal 'id'
      + expected - actual

      -idmax
      +id
```

Reply:
This hack also helps the test pass, but not sure what the actual fix here is:
```typescript
export const aggregationStateInitializer: GridStateInitializer<
  Pick<DataGridPremiumProcessedProps, 'aggregationModel' | 'initialState'>,
  GridPrivateApiPremium
> = (state, props, apiRef) => {
  apiRef.current.caches.aggregation = {
    rulesOnLastColumnHydration: props.aggregationModel !== undefined ? ({ _init: {} } as any) : {},
    rulesOnLastRowHydration: {},
  };

  return {
    ...state,
    aggregation: {
      model: props.aggregationModel ?? props.initialState?.aggregation?.model ?? {},
    },
  };
};
```

Reply:
aha, I think I see the actual fix now

Reply:
I think the right place to update `apiRef.current.caches.aggregation.rulesOnLastColumnHydration` is in the preprocessor after `hydrateColumns` and not in `checkAggregationRulesDiff`